### PR TITLE
keepaway state for AttackerTactic

### LINKED ISSUE #2073: Design a new KeepAwayTactic
### Description of the task
Make a new keep away tactic to dribble the ball away from approaching enemy threats and potentially towards better positioning. Some additional considerations:
* Integration into the [AttackerTactic](https://github.com/UBC-Thunderbots/Software/pull/2071)
* 1m dribble range 

### Acceptance criteria

- [ ] Make a KeepAwayFSM and KeepAwayTactic
- [ ] Test it with FSM tests and Simulated tests

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

- Always pass the best pass so far down to AttackerTactic and have a separate bool to commit to the pass
- When in keepaway state, optimize `ratePassEnemyRisk` to find a nearby point to dribble to
- Simulated tests for keepaway state

The notebook does not really need review, that was purely to iterate on the rest of the code in the PR. 

### Testing Done

Simulated tests for the keepaway state in the attacker FSM
Unit tests for `findKeepAwayTargetPoint` and `ratePasserPointForKeepAway` 

### Resolved Issues

Resolves #2073 

### Length Justification

Unfortunately this ended up being a chungus of a pull request with a bit of scope creep, but I did make a copy of the `ratepass_plot.py` notebook to hack on this, that does not really need review. Another substantial chunk of this PR is tests. 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: commenting for later, this should be `false`

File: src/software/ai/hl/stp/tactic/attacker/attacker_tactic.h
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/attacker/attacker_tactic.h]: merge conflict stuff, fix this

File: src/python_tools/notebooks/plot_enemy_risk_passer_point.py
Comment (jonathanlew) [inline src/python_tools/notebooks/plot_enemy_risk_passer_point.py]: use MM_PER_M?

File: src/software/ai/evaluation/keep_away.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.cpp:15]: it seems like they could be tuned, so should these be dynamic parameters?

File: src/software/ai/evaluation/keep_away.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.cpp]: leftover todo?

File: src/software/ai/evaluation/keep_away.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.cpp]: can you measure how long these two iterations take and whether it causes a significant (>1ms) change in tick time?

File: src/software/ai/evaluation/keep_away.h
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.h:23]: I'm wondering how well keep away transitions into passing. Can we have a test for that in shootorpassplaytest, for example?

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: weird change

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: todo to revisit

**Review by jonathanlew — COMMENTED** [review]
Cool to see some keep away action! Just a few comments

**Review by akhilveeraghanta — COMMENTED** [review]
nothing to add beyond what has already been said, nice work as always :100:

File: src/software/ai/evaluation/keep_away.cpp
Comment (jordanzliu) [reply src/software/ai/evaluation/keep_away.cpp:15]: It would be a pretty big change to dependency inject `PassingConfig` down to here, and it's very easy to tune these parameters quickly with simulated tests, the notebook, and Python bindings. I discussed this with @akhilveeraghanta at some point, he can also chime in on this

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jordanzliu) [reply src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: merge conflict resolution issues, will fix

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jordanzliu) [reply src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: This is out of scope for the PR, will remove TODO

File: src/software/ai/evaluation/keep_away.cpp
Comment (jordanzliu) [reply src/software/ai/evaluation/keep_away.cpp]: the entire `findKeepAwayTargetPoint` function takes around 100 microseconds, so this is negligible

File: src/software/ai/evaluation/keep_away.h
Comment (jordanzliu) [reply src/software/ai/evaluation/keep_away.h:23]: https://user-images.githubusercontent.com/41462969/121764265-8e6a5c00-caf7-11eb-8658-7f14585ccb44.mp4

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jordanzliu) [reply src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: revisited

File: src/software/ai/evaluation/keep_away.cpp
Comment (jordanzliu) [reply src/software/ai/evaluation/keep_away.cpp]: this was already tweaked, removed todo

**Comment by jared-paul** [comment]
noticed this last night: right now it seems that when the robot finds a pass/shot it will pivot to the pass/shot orientation no matter what. Can we add a check to see if there is an enemy robot at that orientation so it doesn't just pivot into the enemy robot.
 
image
image

**Comment by jordanzliu** [comment]
https://github.com/UBC-Thunderbots/Software/issues/2136

**Comment by jared-paul** [comment]
can we add a flag to disable keepaway? currently messes with the freekick play and corner kick play

**Comment by jordanzliu** [comment]
@jared-paul on this branch, it looks like `AttackerTactic` is only yielded when corner kick/free kick plays have committed to a pass. It's weird to me that this would be necessary, can you send a video or a replay?

**Comment by jared-paul** [comment]
> @jared-paul on this branch, it looks like `AttackerTactic` is only yielded when corner kick/free kick plays have committed to a pass. It's weird to me that this would be necessary, can you send a video or a replay?

Without keepaway:

https://user-images.githubusercontent.com/17956580/121863587-dda7bc80-ccb0-11eb-9e8d-d62fe280752a.mp4

With keepaway:

https://user-images.githubusercontent.com/17956580/121863627-e6988e00-ccb0-11eb-90e5-5f0cd1043567.mp4

**Comment by jared-paul** [comment]
Nevermind for now. Running them again I think freekickplay is more broken than I thought

File: src/software/ai/evaluation/keep_away.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.cpp:52]: can we please have a simulated test for this case?

Comment (jordanzliu) [reply src/software/ai/evaluation/keep_away.cpp:52]: done

Comment (jonathanlew) [reply src/software/ai/evaluation/keep_away.cpp:52]: Well, I meant a simulated test, i.e. https://drive.google.com/file/d/1mBpdEl8Lo85H85xXXMaUD4cLGH4YnxwQ/view?usp=sharing. In this video, the robot still leaves the field with the ball

**Review by jonathanlew — COMMENTED** [review]
See comment above on field boundary issue

**Comment by jordanzliu** [comment]
@jonathanlew added a simulated test

https://user-images.githubusercontent.com/41462969/122516319-75fcb480-cfc3-11eb-85a6-cabdd8d8b8f4.mp4

File: src/software/ai/hl/stp/tactic/attacker/simulated_attacker_tactic_keep_away_test.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/attacker/simulated_attacker_tactic_keep_away_test.cpp:196]: in this test, it looks like keepaway is direct the robot towards enemy robot 3 even though there's empty space in the -y direction

Comment (jordanzliu) [reply src/software/ai/hl/stp/tactic/attacker/simulated_attacker_tactic_keep_away_test.cpp:196]: did some parameter changes for marginally better perofrmance on this test case

File: src/software/ai/evaluation/keep_away.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.cpp:12]: @akhilveeraghanta thoughts on this?

File: src/software/ai/evaluation/keep_away.cpp
Comment (jonathanlew) [inline src/software/ai/evaluation/keep_away.cpp:34]: wondering if any constants in this file need to scale with div a field

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: the todo is still there

**Review by jonathanlew — COMMENTED** [review]
just a few comments, the simulated test looks great!

File: src/software/ai/evaluation/keep_away.cpp
Comment (jordanzliu) [reply src/software/ai/evaluation/keep_away.cpp:34]: No, this would only possibly need to increase if we wanted to search a larger area, which would be motivated by higher acceleration/speed limit

File: src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h
Comment (jordanzliu) [reply src/software/ai/hl/stp/tactic/attacker/attacker_fsm.h]: deleted the todo

File: src/software/ai/evaluation/keep_away.cpp
Comment (akhilveeraghanta) [reply src/software/ai/evaluation/keep_away.cpp:12]: yeah this requires a full RFC, TODO is fine though, if we chose to stay with gradient descent we need to "tune" anyway

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm! :100: :fire:

**Review by jonathanlew — DISMISSED** [review]
lgtm! I'd suggest merging in upstream/master first before merging though :)

**Review by jonathanlew — APPROVED** [review]
LGTM pending ci

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**
